### PR TITLE
Internal ITrace: Adds back Line Numbers

### PR DESCRIPTION
**Pull request by bchong95**
# Internal ITrace: Adds back Line Numbers

## Description

Adds back the line numbers. Uses a cache for the name.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.cs [hunk 1]
```diff
@@ -56,5 +60,16 @@ public enum AsciiType
             ClassicRounded,
             ExclamationMarks,
         }
+
+        public static string GetFileNameFromPath(string filePath)
+        {
+            if (!FilePathToName.TryGetValue(filePath, out string fileName))
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.cs:66]: Is there any reason to not cache the entire caller info content? That way it avoids building the entire string each time.

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.cs:66]: FilePath is a member reference so no string is being built. Notice that name is only being built on a cache miss.

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.cs:66]: Also cacheing the whole caller info would involve writing a hash function and also we would not leverage the same file name for traces that comes from the same file but different line numbers.

File: Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs [hunk 2]
```diff
@@ -39,21 +39,19 @@ private static class TraceJsonWriter
...
                 writer.WriteFieldName("caller information");
                 writer.WriteObjectStart();
 
                 writer.WriteFieldName("member name");
                 writer.WriteStringValue(trace.CallerInfo.MemberName);
 
-                writer.WriteFieldName("file path");
-                writer.WriteStringValue(trace.CallerInfo.FilePath);
+                writer.WriteFieldName("file name");
+                writer.WriteStringValue(GetFileNameFromPath(trace.CallerInfo.FilePath));
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs:49]: To reduce the output size should this just be `"caller info": "MemberName@FilePath:42"` and follow the text output?

File: Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.cs [hunk 3]
```diff
@@ -56,5 +60,16 @@ public enum AsciiType
...
+            {
+                fileName = filePath.Split('\\').Last();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.cs:68]: Is dictionary faster than doing the following?
int splitPos = filePath.LastIndexOf('\\');
fileName = filePath.SubString(splitPos, filePath.Length);

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.cs:68]: This approach might be faster than the hashing, but it would create an allocation on every filepath, which get reused quite a bit (we can enumerate them in source code).

File: Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs (hunk 2 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs:49]: text formats are hard to parse, which goes against our design goals for JSON. It's probably a 20% bloat in text size.

File: Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs [hunk 4]
```diff
@@ -39,21 +39,19 @@ private static class TraceJsonWriter
...
-#if INTERNAL
                 writer.WriteFieldName("caller information");
                 writer.WriteObjectStart();
 
                 writer.WriteFieldName("member name");
                 writer.WriteStringValue(trace.CallerInfo.MemberName);
 
-                writer.WriteFieldName("file path");
-                writer.WriteStringValue(trace.CallerInfo.FilePath);
+                writer.WriteFieldName("file name");
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs:48]: Can we try to reduce the size of the output by removing making the naming more compact?
```json
"caller info": {
   "member":"",
   "file":"",
   "line":124,
}
```

File: Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs [hunk 5]
```diff
@@ -39,21 +39,19 @@ private static class TraceJsonWriter
                 writer.WriteFieldName("component");
                 writer.WriteStringValue(trace.Component.ToString());
 
-#if INTERNAL
                 writer.WriteFieldName("caller information");
                 writer.WriteObjectStart();
 
                 writer.WriteFieldName("member name");
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs:45]: What is the use case for needing the member name? It already has the file and line number.

Comment (FabianMeiswinkel) [reply Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs:45]: IMO having the member name is beneficial if anyone isn't able to map directly with source file of the absolutely correct release version. SO I would keep it unless there is a significant overhead (which seems very unlikely)

File: Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.cs (hunk 3 above)
Comment (FabianMeiswinkel) [reply Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.cs:68]: Same operations exist on ReadOnlySpan as well - which would avoid allocation - but I doubt that it would be faster than Dictionary lookup

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - more compact naming would be a useful change IMO

File: Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.cs (hunk 3 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.cs:68]: I thought of that and the problem is that newtonsoft doesn't have support for span from what I could see which means it would do the allocation.

File: Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs (hunk 4 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs]: The bulk of the bloat is in the value and not the names, but ill add it.

**Review by sboshra — APPROVED** [review]
